### ISSUE TITLE
Adding the possibility to run multiple PiwikTracker in one application.

### ISSUE THREAD

**Issue by brototyp** [matomo-org/matomo-sdk-ios#118]
Do we want to support multiple PiwikTracker instances?

**Comment by minitrue22** [comment]
I think this is a good design goal. 
Stay away from those singletons (and some other stuff) and it should be fine.

**Comment by brototyp** [comment]
I was just thinking about this. I don't think it is necessary to change the PiwikUserDefaults for the most part. Most of the values are user specific (number of visits, last visit date and such). But there are a few there I think it would be good to store different values for different Tracker (Backends). I am thinking about the clientId and the optOut. Both might different for different backends. What do you think?

**Comment by elitalon** [comment]
The discussion here seems to be a bit stale, so I apologise in advance if I'm out of context.

Is there any reason to have multiple trackers? I mean, do we have enough evidence that this is something clients actually need? Can we state what benefits it will bring?

I think having different instances might lead to potential errors if clients misconfigure the PiwikTracker instances. For instance:

* Having two or more trackers with the same ID might end up tracking duplicated events.
* Multiple trackers would force developers to maintain a mental model of "which tracker tracks what".

I know, at the end of the day it's the responsibility of library clients to make their choices. But I think in this case having a singleton "root" object is a valid use of the pattern.

Anyway, if this definitely needs to be implemented, I'd rather explore other alternatives.

**Comment by brototyp** [comment]
That's true. The discussion moved on in the pull request here #164 (though it is a bit stale there as well).

About the question: I think your point, barely anyone will need multiple tracker, is correct. But I think there are more upsides to the proposed changes. In the core, not depending on singletons (but offering one as a convenience) helps to write cleaner and better testable code.

**Comment by elitalon** [comment]
Yes, internally the library should strive for having only appropriate use cases for the singleton pattern.

My point is more about access control for client needs. While the internal architecture of the library should not rely on singletons, from a user perspective I think it's a valid case to only expose a single entry point:

```swift
public final class PiwikTracker {

  public static let shared = PiwikTracker()

  internal init() {
    // configure a tracker with sensible defaults
  }

}

The simplified example above illustrates my point. For testing and the overall design of the architecture it is possible to declare as many trackers as needed using the internal initialiser. However, for clients it would enough to have access to the singleton `shared` instance.

Just my two cents to avoid exposing more functionality than clients really need.

**Comment by brototyp** [comment]
I must merged #164 which removes the shared instance and enables everyone to use more than one backend to track. It will be released in version 5.0
